src/replication.c

```diff
@@ -1754,7 +1758,14 @@ void updateSlavesWaitingBgsave(int bgsaveerr, int type) {
         /* We can get here via freeClient()->killRDBChild()->checkChildrenDone(). skip disconnected slaves. */
         if (!slave->conn) continue;
 
-        if (slave->replstate == SLAVE_STATE_WAIT_BGSAVE_END) {
+        if (slave->replstate == SLAVE_STATE_SEND_BULK_AND_STREAM) {
+            /* This is the main channel of the slave that received the RDB.
+             * Put it online if RDB delivery is successful. */
+            if (bgsaveerr == C_OK)
+                replicaPutOnline(slave);
```

Comment:
i have a question, for rdb channel replication, for view of master, the replication offset of replica is much bigger than replica's perspective(replica is applying the replication buffer), right? if we put it online, will something bad happen?

Reply:
Hmm, good question! While accumulating, replica does not send "ack"s. From master's POV, this should be same as non-rdbchannel replication. Previously, master was putting rdb bytes and then replication stream bytes to the network without waiting acks. Same thing happening with two connections this time. I feel like it should be okay.

Reply:
From POV of master, the offset of replication is ack offset, i missed that, so what i worried makes no sense.

BTW, for the stream connection in rdbchannel replication, does this connection have a keepalive mechanism? i found we send `\n` during loading accumulated buffer, how does it work during receiving replication buffer

Reply:
Currently, master checks keepalive only when slave becomes online, see https://github.com/redis/redis/blob/057f039c4b762b3721c2aa2beb95838bd529313e/src/replication.c#L4595

It was already like that before rdbchannel repl, so rdbchannel repl inherited this behavior. During delivery of rdb and repl stream, we don't have keepalive from POV of master. Master will put replica online when bgsave is done and replica will start sending newline while loading rdb and streaming repl buffer. Maybe this is something we need to improve sometime and have keepalive during this period.

Reply:
Thank you. And i found during delivery of rdb in the background, we set timeout for the connection
https://github.com/redis/redis/blob/057f039c4b762b3721c2aa2beb95838bd529313e/src/rdb.c#L3929
so i think there is no big damage, the replica that's hung can not prevent the fork child from terminating

Reply:
yes, rdb fork will eventually realize something is wrong in that case.